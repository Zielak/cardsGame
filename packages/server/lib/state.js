"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var colyseus_1 = require("colyseus");
var eventemitter3_1 = require("eventemitter3");
var logs_1 = require("./logs");
var utils_1 = require("@cardsgame/utils");
var entity_1 = require("./entity");
var entityMap_1 = require("./entityMap");
var State = /** @class */ (function (_super) {
    __extends(State, _super);
    function State(options) {
        var _this = _super.call(this) || this;
        // 60 cm
        _this.tableWidth = 600;
        // 60 cm
        _this.tableHeight = 600;
        _this.players = new entityMap_1.EntityMap();
        _this.isGameStarted = false;
        _this._lastID = -1;
        _this._allEntities = new Map();
        // TODO: do something with these options.
        _this.entities = new entity_1.Entity({
            state: _this,
            type: "root",
            name: "root"
        });
        _this.setupListeners();
        return _this;
    }
    Object.defineProperty(State.prototype, "currentPlayer", {
        get: function () {
            return this.players[this.currentPlayerIdx];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(State.prototype, "playersCount", {
        get: function () {
            return this.players.length;
        },
        enumerable: true,
        configurable: true
    });
    State.prototype.setupListeners = function () {
        var _this = this;
        this.on(utils_1.StateEvents.privatePropsSyncRequest, function (client) {
            // Bubble it down to every entity
            logs_1.logs.info("State.privatePropsSyncRequest");
            _this.entities.emit(utils_1.StateEvents.privatePropsSyncRequest, client);
        });
    };
    /**
     * FIXME: that should be client-side code...
     * Player playing the game needs to have his cards visible
     * at HIS pottom of the screen...
     */
    State.prototype.updatePlayers = function () {
        var _this = this;
        this.entities.childrenArray
            .filter(function (entity) { return entity.type === "player"; })
            .forEach(function (player, idx, players) {
            var angle = ((Math.PI * 2) / players.length) * idx;
            var point = {
                x: Math.sin(angle) * (_this.tableWidth * 0.4),
                y: Math.cos(angle) * (_this.tableHeight * 0.4)
            };
            player.angle = -utils_1.rad2deg(angle);
            player.x = point.x;
            player.y = point.y;
            logs_1.logs.log("updatePlayers", "player[" + idx + "] angle: " + angle + ", point: (" + point.x + ", " + point.y + ")");
        });
    };
    /**
     * Registers new entity to the gamestate
     * @param entity
     * @returns new ID to be assigned to that entity
     */
    State.prototype.rememberEntity = function (entity) {
        var newID = ++this._lastID;
        this._allEntities.set(newID, entity);
        return newID;
    };
    State.prototype.getEntity = function (idOrPath) {
        if (Array.isArray(idOrPath)) {
            var travel_1 = function (entity, path) {
                var idx = path.shift();
                var newChild = entity.children[idx];
                if (!newChild) {
                    logs_1.logs.error("getEntity/path", "This entity doesn't have such child");
                }
                if (path.length > 0) {
                    return travel_1(newChild, path);
                }
                else {
                    return newChild;
                }
            };
            return travel_1(this.entities, __spread(idOrPath));
        }
        return this._allEntities.get(idOrPath);
    };
    /**
     * Gets an array of all entities from the top-most parent
     * to the lowest of the child.
     */
    State.prototype.getEntitiesAlongPath = function (path) {
        var travel = function (entity, path, result) {
            if (result === void 0) { result = []; }
            var idx = path.shift();
            var newChild = entity.children[idx];
            if (!newChild) {
                logs_1.logs.error("getEntity/path", "This entity doesn't have such child");
            }
            result.push(newChild);
            if (path.length > 0) {
                return travel(newChild, path, result);
            }
            else {
                return result;
            }
        };
        return travel(this.entities, __spread(path));
    };
    State.prototype.logTreeState = function () {
        logs_1.logs.log("");
        var indent = function (level) {
            return "│ ".repeat(level);
        };
        var travel = function (entity, level) {
            if (level === void 0) { level = 0; }
            entity.childrenArray.map(function (child, idx, entities) {
                if (child.parentEntity.isContainer &&
                    entities.length > 5 &&
                    idx < entities.length - 5) {
                    // That's too much, man!
                    if (idx === 0) {
                        logs_1.logs.log("" + indent(level), "...");
                    }
                    return;
                }
                var owner = child.owner;
                var lastChild = entities.length - 1 === idx;
                var sIdx = idx === child.idx ? "" + idx : "e" + child.idx + ":s" + idx;
                var sVisibility = child.visibleToPublic ? "Pub" : "";
                var sChildren = child.length > 0 ? child.length + " children" : "";
                var sOwner = owner ? "(" + owner.type + " " + owner.clientID + ")" : "";
                var branchSymbol = lastChild ? "┕━" : "┝━";
                logs_1.logs.log("" + indent(level) + branchSymbol + "[" + sIdx + "]" + sVisibility, child.type + ":" + child.name + "-[" + child.idx + "]", sChildren, sOwner);
                if (child.length > 0) {
                    travel(child, level + 1);
                }
            });
        };
        var root = this.entities;
        logs_1.logs.log("\u250D\u2501[" + root.idx + "]", root.type + ":" + root.name, root.length + "children");
        travel(root, 1);
    };
    State.ROOT_ID = 0;
    __decorate([
        colyseus_1.nosync
    ], State.prototype, "_lastID", void 0);
    __decorate([
        colyseus_1.nosync
    ], State.prototype, "_allEntities", void 0);
    return State;
}(eventemitter3_1.EventEmitter));
exports.State = State;
// Get rid of EventEmitter stuff from the client
colyseus_1.nosync(State.prototype, "_events");
colyseus_1.nosync(State.prototype, "_eventsCount");
colyseus_1.nosync(State.prototype, "_maxListeners");
colyseus_1.nosync(State.prototype, "domain");
//# sourceMappingURL=state.js.map