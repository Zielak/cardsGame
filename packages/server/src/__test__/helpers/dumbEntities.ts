import { canBeChild } from "@/annotations/canBeChild.js"
import { containsChildren } from "@/annotations/containsChildren.js"
import { ChildTrait } from "@/traits/child.js"
import { applyTraitsMixins, Entity } from "@/traits/entity.js"
import { IdentityTrait } from "@/traits/identity.js"
import { ParentTrait } from "@/traits/parent.js"

@canBeChild
@containsChildren
@applyTraitsMixins([IdentityTrait, ParentTrait, ChildTrait])
export class DumbParent extends Entity<DumbParent> {}
export interface DumbParent extends IdentityTrait, ParentTrait, ChildTrait {}

@canBeChild
@applyTraitsMixins([IdentityTrait, ChildTrait])
export class DumbEntity extends Entity<DumbEntity> {}
export interface DumbEntity extends IdentityTrait, ChildTrait {}
