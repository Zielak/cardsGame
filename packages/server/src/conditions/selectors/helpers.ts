import { IEntity } from "../../traits/entity"
import { logs } from "../../logs"

export const propsMatch = (propName: string, values: any[]) => (
  entity: IEntity
) => {
  const result = values.some(value => entity[propName] === value)
  if (!result) {
    logs.warn(
      `propsMatch ${propName}`,
      `entity[${propName}] doesn't match any accepted values:`,
      values
    )
  }
  return result
}
