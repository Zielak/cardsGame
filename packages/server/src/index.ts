export { Server } from "colyseus"
export { RoomConstructor } from "colyseus/lib/Room"
export { Schema, ArraySchema, MapSchema } from "@colyseus/schema"

export { canBeChild } from "./annotations/canBeChild"
export { containsChildren } from "./annotations/containsChildren"
export { defineTypes, type } from "./annotations/type"

export * from "./bots/botNeuron"

import * as commands from "./commands"

export { commands }

export * from "./command"

export * from "./room"
export * from "./state/state"
export * from "./state/helpers"

export * from "./actionTemplate"

export { Conditions } from "./conditions"
export { getFlag, setFlag } from "./conditions/utils"
export { BotConditions, EntityConditions } from "./bots/conditions"
export { ClientEventConditions } from "./interaction"

export { testAction } from "./interaction"

export { Player, ServerPlayerEvent } from "./players/player"
export { Bot } from "./players/bot"

export * from "./traits/boxModel"
export * from "./traits/child"
export * from "./traits/entity"
export * from "./traits/flexyContainer"
export * from "./traits/identity"
export * from "./traits/label"
export * from "./traits/location"
export * from "./traits/ownership"
export * from "./traits/parent"
export * from "./traits/parentArray"
export * from "./traits/parentMap"
export * from "./traits/selectableChildren"
export * from "./traits/twoSided"

export * from "./entities"

export { populatePlayerEvent } from "./utils"
