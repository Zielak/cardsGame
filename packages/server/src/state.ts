import { nosync } from "colyseus"
import { EventEmitter } from "eventemitter3"
import { logs } from "./logs"
import { deg2rad, rad2deg, StateEvents } from "@cardsgame/utils"
import { Entity } from "./entity"
import { EntityMap } from "./entityMap"
import { Player } from "./player"

export class State extends EventEmitter {
  // 60 cm
  tableWidth = 600
  // 60 cm
  tableHeight = 600

  entities: Entity

  players = new EntityMap<PlayerData>()
  clients = new EntityMap<string>()
  currentPlayerIdx: number
  get currentPlayer(): PlayerData {
    return this.players[this.currentPlayerIdx]
  }
  get playersCount(): number {
    return this.players.length
  }

  // TODO: think this through:
  gameVariants: any

  isGameStarted = false
  @nosync
  _lastID = -1
  @nosync
  _allEntities = new Map<number, Entity>()

  constructor(options?: IStateOptions) {
    super()
    // TODO: do something with these options.

    this.entities = new Entity({
      state: this,
      type: "root",
      name: "root"
    })

    this.setupListeners()
  }

  setupListeners() {
    this.on(StateEvents.privatePropsSyncRequest, (client: string) => {
      // Bubble it down to every entity
      logs.info("State.privatePropsSyncRequest")
      this.entities.emit(StateEvents.privatePropsSyncRequest, client)
    })
  }

  /**
   * Registers new entity to the gamestate
   * @param entity
   * @returns new ID to be assigned to that entity
   */
  rememberEntity(entity: Entity) {
    const newID = ++this._lastID
    this._allEntities.set(newID, entity)
    return newID
  }

  /**
   * Get an Entity by its ID
   * @param id
   */
  getEntity(id: EntityID): Entity
  /**
   * Get an Entity by its idx path
   * @param path
   */
  getEntity(path: number[]): Entity
  getEntity(idOrPath: EntityID | number[]): Entity {
    if (Array.isArray(idOrPath)) {
      const travel = (entity: Entity, path: number[]) => {
        const idx = path.shift()
        const newChild = entity.children[idx] as Entity
        if (!newChild) {
          logs.error("getEntity/path", `This entity doesn't have such child`)
        }
        if (path.length > 0) {
          return travel(newChild, path)
        } else {
          return newChild
        }
      }
      return travel(this.entities, [...idOrPath])
    }
    return this._allEntities.get(idOrPath)
  }

  /**
   * Gets an array of all entities from the top-most parent
   * to the lowest of the child.
   */
  getEntitiesAlongPath(path: number[]): Entity[] {
    const travel = (entity: Entity, path: number[], result: Entity[] = []) => {
      const idx = path.shift()
      const newChild = entity.children[idx] as Entity
      if (!newChild) {
        logs.error("getEntity/path", `This entity doesn't have such child`)
      }
      result.push(newChild)
      if (path.length > 0) {
        return travel(newChild, path, result)
      } else {
        return result
      }
    }
    return travel(this.entities, [...path])
  }

  logTreeState() {
    logs.log("")
    const indent = (level: number) => {
      return "│ ".repeat(level)
    }
    const travel = (entity: Entity, level: number = 0) => {
      entity.childrenArray.map((child, idx, entities) => {
        if (
          child.parentEntity.isContainer &&
          entities.length > 5 &&
          idx < entities.length - 5
        ) {
          // That's too much, man!
          if (idx === 0) {
            logs.log(`${indent(level)}`, "...")
          }
          return
        }

        const owner = child.owner

        const lastChild = entities.length - 1 === idx
        const sIdx = idx === child.idx ? `${idx}` : `e${child.idx}:s${idx}`
        const sVisibility = child.visibleToPublic ? "Pub" : ""

        const sChildren = child.length > 0 ? child.length + " children" : ""
        const sOwner = owner ? `(${owner.type} ${owner.clientID})` : ""
        const branchSymbol = lastChild ? "┕━" : "┝━"

        logs.log(
          `${indent(level)}${branchSymbol}[${sIdx}]${sVisibility}`,
          `${child.type}:${child.name}-[${child.idx}]`,
          sChildren,
          sOwner
        )
        if (child.length > 0) {
          travel(child, level + 1)
        }
      })
    }
    const root = this.entities
    logs.log(
      `┍━[${root.idx}]`,
      `${root.type}:${root.name}`,
      root.length + "children"
    )
    travel(root, 1)
  }

  static ROOT_ID = 0
}

// Get rid of EventEmitter stuff from the client
nosync(State.prototype, "_events")
nosync(State.prototype, "_eventsCount")
nosync(State.prototype, "_maxListeners")
nosync(State.prototype, "domain")

export interface IStateOptions {
  minClients: number
  maxClients: number
  hostID: string
}

export type PlayerData = {
  entity: Entity
  clientID: string
}
