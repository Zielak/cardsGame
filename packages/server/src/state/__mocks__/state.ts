import type { Player } from "../../player/player.js"

export class State<
  V extends Record<string, unknown> = Record<string, unknown>
> {
  name = "Unnamed"
  type = "state"
  tableWidth = 60 // 60 cm
  tableHeight = 60 // 60 cm
  clients = new Array<string>()
  turnBased = true
  round = 0
  players = new Array<Player>()
  currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }
  isGameStarted = false
  isGameOver = false

  ui = {}

  playerViewPosition = {
    alignX: "center",
    alignY: "bottom",
    paddingX: 0,
    paddingY: 0,
  }

  variantData: V
  hijacksInteractionTarget = false

  id = 0

  _registerEntity(): void {
    // do nothing
  }
}
