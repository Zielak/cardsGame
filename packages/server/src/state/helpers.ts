import { limit, logs } from "@cardsgame/utils"

import { Bot, isBot } from "../players/bot"
import { Player } from "../players/player"
import { IdentityTrait } from "../traits/identity"
import { hasChildren, isParent, ParentTrait } from "../traits/parent"

import { State } from "."

/**
 * Will get you an index of given player in turn queue.
 * Useful if you happen to have just a `Player` reference at hand.
 */
export function getPlayersIndex(state: State, player: Player): number {
  return state.players.findIndex((p) => p === player)
}

export function getNextPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current + 1 === state.players.length ? 0 : current + 1
}

export function getNextPlayer(state: State): Player {
  return state.players[getNextPlayerIdx(state)]
}

export function getPreviousPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current - 1 === -1 ? state.players.length - 1 : current - 1
}

export function getPreviousPlayer(state: State): Player {
  return state.players[getPreviousPlayerIdx(state)]
}

export function getPlayerByName(state: State, name: string): Player {
  return state.players.find((player) => player.name === name)
}

export function getAllBots(state: State): Bot[] {
  return state.players.filter((player) => isBot(player)) as Bot[]
}

/**
 * Gets an array of all entities from the top-most parent
 * to the lowest of the child.
 */
export function getEntitiesAlongPath(
  state: State,
  path: number[]
): IdentityTrait[] {
  const travel = (
    parent: ParentTrait,
    remainingPath: number[],
    result: any[] = []
  ): any[] => {
    const idx = remainingPath.shift()
    const newChild = parent.getChild(idx)
    if (!newChild) {
      logs.error("getEntitiesAlongPath", `This entity doesn't have such child.`)
      return result
    }

    if (remainingPath.length > 0 && !hasChildren(newChild)) {
      logs.error(
        "getEntitiesAlongPath",
        `Path inaccessible, entity doesn't have any children. Stopped at [${path}]. Remaining path: [${remainingPath}]`
      )
      return result
    }

    result.push(newChild)
    if (remainingPath.length > 0 && isParent(newChild)) {
      return travel(newChild, remainingPath, result)
    } else {
      return result
    }
  }
  return travel(state, [...path])
}
