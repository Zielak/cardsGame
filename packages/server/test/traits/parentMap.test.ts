import { DumbEntity, DumbMapParent } from "../helpers/dumbEntities"
import { IdentityTrait } from "../../src/traits/identity"
import { ChildTrait } from "../../src/traits/child"
import { State } from "../../src/state"

let state: State

beforeEach(() => {
  state = new State()
})

describe("ParentConstructor", () => {
  test("sets default properties", () => {
    const parent = new DumbMapParent(state)

    expect(parent.countChildren()).toBe(0)

    expect(parent.hijacksInteractionTarget).toBeDefined()
    expect(parent.hijacksInteractionTarget).toBe(true)

    expect(parent.maxChildren).toBeDefined()
    expect(parent.maxChildren).toBe(Infinity)
  })
  test("remembers custom values", () => {
    const parent = new DumbMapParent(state, { maxChildren: 3 })

    expect(parent.maxChildren).toBe(3)
  })
})

describe("#removeChildAt", () => {
  let parent: DumbMapParent
  beforeEach(() => {
    parent = new DumbMapParent(state)
  })
  it(`throws on invalid "idx"`, () => {
    expect(parent.removeChildAt(-2)).toBe(false)
    expect(parent.removeChildAt(NaN)).toBe(false)
    expect(parent.removeChildAt(Infinity)).toBe(false)
    expect(parent.removeChildAt(10)).toBe(false)
  })
  it("removes the only child", () => {
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChildAt(0)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })
  it("removes child in the middle", () => {
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(3)
    expect(parent.removeChildAt(1)).toBe(true)
    expect(parent.countChildren()).toBe(2)
  })
})

describe("#removeChild", () => {
  let parent: DumbMapParent
  it("calls removeChildAt", () => {
    parent = new DumbMapParent(state)
    const entity = new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChild(entity)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })

  it("finds proper child", () => {
    parent = new DumbMapParent(state)
    new DumbEntity(state, { parent })
    const entity = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    parent.removeChild(entity)
    expect(
      parent.getChildren().map((e: ChildTrait & IdentityTrait) => e.id)
    ).toMatchSnapshot()
  })
})

describe("#addChild", () => {
  let parent: DumbMapParent
  let entity: DumbEntity
  beforeEach(() => {
    parent = new DumbMapParent(state)
    new DumbEntity(state, { parent })
  })
  it("adds new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(1)
    expect(parent.countChildren()).toBe(2)
  })
  it("prepends new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity, true)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(0)
    expect(parent.countChildren()).toBe(2)
  })
  it("prepends in busier parent", () => {
    // Clean parent this time.
    parent = new DumbMapParent(state)
    const e0 = new DumbEntity(state, { parent, idx: 0 })
    const e1 = new DumbEntity(state, { parent, idx: 1 })
    const e3 = new DumbEntity(state, { parent, idx: 3 })
    entity = new DumbEntity(state)

    expect(parent.getChildren().map(c => c.idx)).toStrictEqual([0, 1, 3])

    parent.addChild(entity, true)

    expect(entity.parent).toBe(parent)
    expect(parent.getChildren().map(c => c.idx)).toStrictEqual([0, 1, 2, 3])
    expect(entity.idx).toBe(0)
    expect(e0.idx).toBe(1)
    expect(e1.idx).toBe(2)
    expect(e3.idx).toBe(3)

    expect(parent.countChildren()).toBe(4)
  })
})

describe("#moveChildTo", () => {
  const mapMapChildren = (parent: DumbMapParent) => {
    const max =
      parent.maxChildren !== Infinity
        ? parent.maxChildren
        : parent.countChildren()

    const results = []
    for (let i = 0; i < max; i++) {
      const child = parent.getChild<DumbEntity>(i)
      results.push(child ? child.id : null)
    }
    return results
  }

  let parent: DumbMapParent

  describe("filled", () => {
    beforeEach(() => {
      parent = new DumbMapParent(state, { maxChildren: 5 })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
    })
    it("moves item3 down 1 position", () => {
      parent.moveChildTo(2, 1)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 up 1 position", () => {
      parent.moveChildTo(2, 3)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 to bottom", () => {
      parent.moveChildTo(2, 0)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 to top", () => {
      parent.moveChildTo(2, parent.countChildren() - 1)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("doesn't move anything (item3 to same position)", () => {
      parent.moveChildTo(2, 2)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("handles too high 'to' number, treats as max value", () => {
      expect(() => parent.moveChildTo(2, 10)).not.toThrow()
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
  })

  describe("with empty spaces", () => {
    beforeEach(() => {
      parent = new DumbMapParent(state, { maxChildren: 5 })
      new DumbEntity(state, { parent, idx: 2 })
    })
    it("moves item down 1 position", () => {
      parent.moveChildTo(2, 1)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item up 1 position", () => {
      parent.moveChildTo(2, 3)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item to bottom", () => {
      parent.moveChildTo(2, 0)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("moves item to top", () => {
      parent.moveChildTo(2, parent.maxChildren - 1)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("doesn't move anything (item to same position)", () => {
      parent.moveChildTo(2, 2)
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
    it("handles too high 'to' number, treats as max value", () => {
      expect(() => parent.moveChildTo(2, 10)).not.toThrow()
      expect(mapMapChildren(parent)).toMatchSnapshot()
    })
  })
})

test.todo("#countChildren")
describe("#getChildren", () => {
  let parent: DumbMapParent
  beforeEach(() => {
    parent = new DumbMapParent(state)
  })
  test("empty parent", () => {
    expect(parent.getChildren()).toMatchSnapshot()
  })
  test("couple of children", () => {
    const first = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    const last = new DumbEntity(state, { parent })

    const result = parent.getChildren()
    expect(result).toMatchSnapshot()
    expect(result[0]).toBe(first)
    expect(result[result.length - 1]).toBe(last)
  })
})

describe("#getChild", () => {
  let parent
  let entity

  it("gets child", () => {
    parent = new DumbMapParent(state)

    new DumbEntity(state, { parent })
    entity = new DumbEntity(state, { parent })

    expect(parent.getChild(1)).toBe(entity)
  })
  test("idx out of range", () => {
    parent = new DumbMapParent(state)

    expect(parent.getChild(-1)).toBeUndefined()
    expect(parent.getChild(0)).toBeUndefined()
    expect(parent.getChild(Infinity)).toBeUndefined()
  })
  test("idx out of range in limited container", () => {
    parent = new DumbMapParent(state, { maxChildren: 5 })

    expect(parent.getChild(-1)).toBeUndefined()
    expect(parent.getChild(10)).toBeUndefined()
    expect(parent.getChild(-Infinity)).toBeUndefined()
    expect(parent.getChild(Infinity)).toBeUndefined()
  })
})

test("#getTop", () => {
  const parent = new DumbMapParent(state)
  new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })
  const entity = new DumbEntity(state, { parent })

  expect(parent.getTop()).toBe(entity)
})

test("#getBottom", () => {
  const parent = new DumbMapParent(state)
  const entity = new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })

  expect(parent.getBottom()).toBe(entity)
})

describe("#getFirstEmptySpot", () => {
  let parent: DumbMapParent
  describe("limited parent", () => {
    beforeEach(() => {
      parent = new DumbMapParent(state, { maxChildren: 3 })
    })
    it("returns zero on empty parent", () => {
      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return zero", () => {
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return one", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(1)
    })
    it("return last", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })

      expect(parent.getFirstEmptySpot()).toBe(2)
    })
    it("can't find it so it returns -1", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(-1)
    })
  })
  describe("unlimited parent", () => {
    beforeEach(() => {
      parent = new DumbMapParent(state)
    })
    it("returns zero on empty parent", () => {
      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return zero", () => {
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return one", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(1)
    })
    it("return last", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })

      expect(parent.getFirstEmptySpot()).toBe(2)
    })
  })
})

describe("#getLastEmptySpot", () => {
  let parent: DumbMapParent
  beforeEach(() => {
    parent = new DumbMapParent(state, { maxChildren: 5 })
  })
  it("picks last on empty parent", () => {
    expect(parent.getLastEmptySpot()).toBe(4)
  })
  it("picks last spot", () => {
    // 0
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    // 3
    // 4

    expect(parent.getLastEmptySpot()).toBe(4)
  })
  it("picks 3 from 2 empty spaces", () => {
    // 0
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    // 3
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(3)
  })
  it("picks 2 from 2 empty spaces", () => {
    new DumbEntity(state, { parent, idx: 0 })
    // 1
    // 2
    new DumbEntity(state, { parent, idx: 3 })
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(2)
  })
  it("returns -1, for filled container", () => {
    new DumbEntity(state, { parent, idx: 0 })
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    new DumbEntity(state, { parent, idx: 3 })
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(-1)
  })
})
