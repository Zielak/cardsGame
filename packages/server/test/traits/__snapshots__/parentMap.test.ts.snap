// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getChildren couple of children 1`] = `
Array [
  Object {
    "idx": 0,
  },
  Object {
    "idx": 1,
  },
  Object {
    "idx": 2,
  },
  Object {
    "idx": 3,
  },
  Object {
    "idx": 4,
  },
]
`;

exports[`#getChildren empty parent 1`] = `Array []`;

exports[`#moveChildTo filled doesn't move anything (item3 to same position) 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`#moveChildTo filled throws on indexes out of boundaries 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 down 1 position 1`] = `
Array [
  1,
  3,
  2,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 to bottom 1`] = `
Array [
  3,
  2,
  1,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 to top 1`] = `
Array [
  1,
  2,
  5,
  4,
  3,
]
`;

exports[`#moveChildTo filled moves item3 up 1 position 1`] = `
Array [
  1,
  2,
  4,
  3,
  5,
]
`;

exports[`#moveChildTo with empty spaces doesn't move anything (item to same position) 1`] = `
Array [
  null,
  null,
  1,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces throws on indexes out of boundaries 1`] = `
Array [
  null,
  null,
  1,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item down 1 position 1`] = `
Array [
  null,
  1,
  null,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item to bottom 1`] = `
Array [
  1,
  null,
  null,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item to top 1`] = `
Array [
  null,
  null,
  null,
  null,
  1,
]
`;

exports[`#moveChildTo with empty spaces moves item up 1 position 1`] = `
Array [
  null,
  null,
  null,
  1,
  null,
]
`;

exports[`#removeChild finds proper child 1`] = `
Array [
  1,
  3,
]
`;
